library/alloc/src/collections/btree/map.rs

```diff
@@ -1947,16 +1947,9 @@ impl<K, V> FusedIterator for RangeMut<'_, K, V> {}
 
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<K: Ord, V> FromIterator<(K, V)> for BTreeMap<K, V> {
-    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> BTreeMap<K, V> {
-        let mut inputs: Vec<_> = iter.into_iter().collect();
```

Comment:
I think you are right that existing codes already optimized the case when passing a vec directly. However, with this PR, it will further optimize the case where the iteration is advanced. 

Please let me know if this further optimization in the PR is desirable. Thanks.

Reply:
Well, it still applies when the vec has been advanced to contain fewer entries than half its capacity.

https://github.com/rust-lang/rust/blob/09064a21b05df8a5f92d34622c467bd037a923fd/library/alloc/src/vec/spec_from_iter.rs#L42-L47

Turning a Vec iterator that is more-than-half-empty into a BtreeMap/Set seems like too rare a case to optimize for.

Reply:
While this PR does avoid one extra data copy regardless of the capacity.

I will close this PR then. Thanks for reviewing the PR.